src/Microsoft.OData.Core/Json/ODataJsonWriterUtils.cs

```diff
@@ -224,27 +224,40 @@ private static void WriteInnerError(IJsonWriter jsonWriter, ODataInnerError inne
             jsonWriter.WriteName(innerErrorPropertyName);
             jsonWriter.StartObjectScope();
 
-            //// NOTE: we add empty elements if no information is provided for the message, error type and stack trace
-            ////       to stay compatible with Astoria.
+            if (innerError.Properties != null)
+            {
+                foreach (KeyValuePair<string, string> pair in innerError.Properties)
+                {
+                    jsonWriter.WriteName(pair.Key);
+                    jsonWriter.WriteValue(pair.Value);
+                }
+            }
+            else
+            {
+                //// NOTE: we add empty elements if no information is provided for the message, error type and stack trace
+                ////       to stay compatible with Astoria.
 
-            // "message": "<message>"
-            jsonWriter.WriteName(JsonConstants.ODataErrorInnerErrorMessageName);
-            jsonWriter.WriteValue(innerError.Message ?? string.Empty);
+                // "message": "<message>"

```

Comment:
>// "message": "<message>" [](start = 16, length = 25)

if we save the "message", "type", "stacktrace" into the Properties, it seems we don't need these codes?

Reply:
Only problem is, the new code could change the order that the properties are written in the JSON payload, which may be considered a breaking change.  Perhaps we should still write the "known" properties first, and then write the remaining properties.

---
In reply to: [315445762](https://github.com/OData/odata.net/pull/1523#discussion_r315445762) [](ancestors = 315445762)